File: backend/app/services/parser.py
import re
from typing import Dict, List, Optional

class ResumeParser:
    """
    Heuristic-based parser to split resume text into sections.
    """
    
    SECTION_HEADERS = {
        "experience": ["experience", "work experience", "employment", "work history", "professional experience"],
        "education": ["education", "academic background", "qualifications", "education & certifications"],
        "skills": ["skills", "technical skills", "competencies", "technologies", "core competencies"],
        "projects": ["projects", "personal projects", "academic projects", "key projects"],
        "certifications": ["certifications", "courses", "licenses"]
    }

    @staticmethod
    def parse_sections(text: str) -> Dict[str, str]:
        """
        Splits raw text into sections based on headers.
        """
        lines = text.split('\n')
        sections = {
            "experience": "",
            "education": "",
            "skills": "",
            "projects": "",
            "other": ""
        }
        
        current_section = "other"
        
        for line in lines:
            line_clean = line.strip().lower()
            
            # Check if line matches a known header
            is_header = False
            for section, keywords in ResumeParser.SECTION_HEADERS.items():
                # specific check: lines that are short and contain keywords
                if len(line_clean) < 50 and any(k == line_clean for k in keywords):
                    current_section = section
                    is_header = True
                    break
                # fuzzy check for headers like "Professional Experience"
                elif len(line_clean) < 50 and any(k in line_clean for k in keywords) and len(line_clean.split()) < 4:
                     # Be careful not to trigger on "I have experience in..."
                     # A header usually doesn't have verbs or full sentences.
                     # This is a basic heuristic.
                     current_section = section
                     is_header = True
                     break
            
            if not is_header:
                if current_section in sections:
                    sections[current_section] += line + "\n"
        
        return sections

File: app/api/lib/parser.ts
/**
 * Resume Parser - Heuristic-based section detection
 * Ported from Python backend
 */

export interface ResumeSections {
    experience: string;
    education: string;
    skills: string;
    projects: string;
    other: string;
}

const SECTION_HEADERS: Record<string, string[]> = {
    experience: ['experience', 'work experience', 'employment', 'work history', 'professional experience'],
    education: ['education', 'academic background', 'qualifications', 'education & certifications'],
    skills: ['skills', 'technical skills', 'competencies', 'technologies', 'core competencies'],
    projects: ['projects', 'personal projects', 'academic projects', 'key projects'],
    certifications: ['certifications', 'courses', 'licenses'],
};

/**
 * Split resume text into sections based on headers
 */
export function parseSections(text: string): ResumeSections {
    const lines = text.split('\n');
    const sections: ResumeSections = {
        experience: '',
        education: '',
        skills: '',
        projects: '',
        other: '',
    };

    let currentSection: keyof ResumeSections = 'other';

    for (const line of lines) {
        const lineClean = line.trim().toLowerCase();

        // Check if line matches a known header
        let isHeader = false;

        for (const [section, keywords] of Object.entries(SECTION_HEADERS)) {
            // Specific check: lines that are short and contain keywords
            if (lineClean.length < 50 && keywords.some((k) => k === lineClean)) {
                // Map certifications to other, otherwise use the section name
                currentSection = section === 'certifications' ? 'other' : (section as keyof ResumeSections);
                isHeader = true;
                break;
            }

            // Fuzzy check for headers like "Professional Experience"
            if (
                lineClean.length < 50 &&
                keywords.some((k) => lineClean.includes(k)) &&
                lineClean.split(/\s+/).length < 4
            ) {
                // Map certifications to other, otherwise use the section name
                currentSection = section === 'certifications' ? 'other' : (section as keyof ResumeSections);
                isHeader = true;
                break;
            }
        }

        if (!isHeader) {
            sections[currentSection] += line + '\n';
        }
    }

    return sections;
}
